### PR TITLE
LibHttpClient Linux: Add Linux build files and configure files to support LibHC Linux.

### PR DESCRIPTION
**Pull request by SebastianPD**
This update is to add LibHC support to Linux. The following code consists of build files, modifications to files, and a new submodule.

This has been tested with WSL and built using Visual Studio 2019.

### CODE REVIEW

**Comment by SahilAshar** [comment]
@lubeltra @jplafonta I've already reviewed this PR privately and it LGTM. Adding y'all in case you want to quickly peruse.

**Comment by ScottMunroMS** [comment]
> @lubeltra @jplafonta I've already reviewed this PR privately and it LGTM. Adding y'all in case you want to quickly peruse.

@SahilAshar @SebastianPD 
It's my understanding that deb's team doesn't own LHC (even if y'all have contributed to it) so I'd like us to hold off on checking in until we get some kind of signoff from their team (even if it's purely a ceremonial rubber stamp)

**Comment by SahilAshar** [comment]
> > @lubeltra @jplafonta I've already reviewed this PR privately and it LGTM. Adding y'all in case you want to quickly peruse.
> 
> @SahilAshar @SebastianPD It's my understanding that deb's team doesn't own LHC (even if y'all have contributed to it) so I'd like us to hold off on checking in until we get some kind of signoff from their team (even if it's purely a ceremonial rubber stamp)

Agreed - I told Seb we'd hold off merging until we got signoffs from other primary contributors.

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (lubeltra) [inline Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:52]: We should avoid platform specific code. Does including this break other platforms?

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (lubeltra) [inline Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:122]: no printf
if you need to log something use the tracing macros

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (lubeltra) [inline Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:104]: why isn't this another implementation of `verify_X509_cert_chain`?

File: Source/HTTP/Curl/CurlEasyRequest.h
Comment (lubeltra) [inline Source/HTTP/Curl/CurlEasyRequest.h:11]: 1) can you expand on what the issue is?
2) can these declarations be moved into cpp files?
3) you should use HC_DECLARE_TRACE_AREA and not name the globals directly

File: Source/HTTP/Curl/CurlEasyRequest.cpp
Comment (lubeltra) [inline Source/HTTP/Curl/CurlEasyRequest.cpp:42]: why?

File: Include/httpClient/config.h
Comment (lubeltra) [inline Include/httpClient/config.h:61]: I would rewrite this as
```c++
#elif defined(__linux__)
    #if defined(__ANDROID)
        #define HC_PLATFORM HC_PLATFORM_ANDROID
    #else
        #define HC_PLATFORM HC_PLATFORM_LINUX
    #endif

    #if defined(__LP64__)
        #define HC_DATAMODEL HC_DATAMODEL_LP64
    #else
        #define HC_DATAMODEL HC_DATAMODEL_ILP32
    #endif
#endif
```

File: Build/libHttpClient.Linux.C/libHttpClient.Linux.C.vcxproj
Comment (lubeltra) [inline Build/libHttpClient.Linux.C/libHttpClient.Linux.C.vcxproj:1]: is this generated from a template?

**Review by lubeltra — CHANGES_REQUESTED** [review]

File: Build/libHttpClient.Linux.C/libHttpClient.Linux.C.vcxproj
Comment (SebastianPD) [reply Build/libHttpClient.Linux.C/libHttpClient.Linux.C.vcxproj:1]: It was generated from creating empty projects in visual studio.

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (SebastianPD) [reply Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:104]: I placed it there due to already having the cert stack vat to be verified with OpenSSL. My previous implementation had me convert it all back to a cert stack in verify_X509_cert_chain but to prevent errors and to not be redundant I leveraged what was already made. From my understanding of the code what's supposed to happen is that openssl should verify before ever reaching here to this file, but it does not, so we step in here at the very end.

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (SebastianPD) [reply Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:52]: I can only test it against windows at the moment, and that works.

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (SebastianPD) [reply Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:122]: this was for debugging. I was supposed to remove this.

File: Source/HTTP/Curl/CurlEasyRequest.h
Comment (SebastianPD) [reply Source/HTTP/Curl/CurlEasyRequest.h:11]: 1. John explained to me what the problem was a while back and it had to do with how the compilers handled the declaration of traces in the .h file. The Clang Compiler doesn't know what should be filled in for g_trace while MSBuild did. 
2. Tried putting them in the cpp files and that did not work.
3. I should do that.

File: Source/HTTP/Curl/CurlEasyRequest.cpp
Comment (SebastianPD) [reply Source/HTTP/Curl/CurlEasyRequest.cpp:42]: It was to resolve a bug I found with the curl library. This doesn't happen in xcurl. What happened was that when creating a WebSocket it would never complete and that was because MLP wanted a websocket it a fixed body. When sending the request for the handshake it would send no data and these setOpts would never be initialized. Removing the if statement solves the issue. Bodysize is unsigned so I we don't have to worry about a negative number.

Comment (ScottMunroMS) [reply Source/HTTP/Curl/CurlEasyRequest.cpp:42]: @SebastianPD do we know that this is actually a bug in curl? or could this just be curl's (perhaps poorly documented) intentional behavior? 

in general, the goal of xcurl is to mirror curl's behavior as closely as reasonably possible. If this isn't definitely a bug it's possible there's reasons xcurl should mirror this as well

> beacuse MLP wanted a websocket it a fixed body
can you elaborate on this? if the bodySize here is zero, what additional data is it now sending?

additionally, have we validated that the requests sent by xcurl haven't changed at all by removing this if-check? even if things still work it'd be interesting to know whether the requests are sent with any additional headers which might not have been there beforeor if the CURLOPT_READFUNCTION is now called in a code path in which it previously wasn't

File: Include/httpClient/config.h
Comment (ScottMunroMS) [inline Include/httpClient/config.h:62]: shouldn't this be `__ANDROID__`?

Comment (SebastianPD) [reply Include/httpClient/config.h:62]: you're right I'll resolve this in my next push. Need to make the templates

File: Source/HTTP/Curl/CurlEasyRequest.cpp
Comment (ScottMunroMS) [reply Source/HTTP/Curl/CurlEasyRequest.cpp:42]: > because MLP wanted a webscoket it a fixed body

I also wonder whether this is a bug in MLP instead of libhttpclient. What requires the fixed body and could be address it at the MLP level?

(if discussing this in any meaningful way requires sharing internal details about MLP let's chat out-of-band, maybe on teams. we should avoid putting MLP internal code in this public repo even in comments)

Comment (SebastianPD) [reply Source/HTTP/Curl/CurlEasyRequest.cpp:42]: what would happen previously when doing the handshake was `Transfer-Encoding: chunked` was a header and removing this if statement and setting the set opt would make the header say it has a size of 0. The body would have nothing in it. Those were the only differences when looking at the fiddler trace.

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (lubeltra) [reply Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:104]: we could pass the cert stack down into `verify_X509_cert_chain`
I really want to minimize the amount of code we ifdef for specific platforms as it really hurts readability

**Review by lubeltra — CHANGES_REQUESTED** [review]

File: Source/WebSocket/Websocketpp/x509_cert_utilities.hpp
Comment (lubeltra) [inline Source/WebSocket/Websocketpp/x509_cert_utilities.hpp:529]: check on HC_PLATFORM

**Review by lubeltra — CHANGES_REQUESTED** [review]

**Review by jasonsandlin — APPROVED** [review]
:shipit:

**Pull request merged**
